config.org
#+TITLE: My Emacs Configuration
#+AUTHOR: Sean Avery
#+EMAIL: seanavery@gmx.net
#+OPTIONS: num:nil
#+STARTUP: overview
#+Tag: [[org:../start-index.org][INDEX]]
#+Setupfile: ~/../../../mnt/c/Users/averh/Dropbox/org-templates/org-html-themes/setup/theme-readtheorg.setup

* Information's
This is my Emacs-Configuration file. Its still in the making.
** Other Options
#+BEGIN_EXAMPLE emacs-lisp
#+HTML_HEAD: <link rel="stylesheet" type="text/css" href="rethink.css" />
#+OPTIONS: toc:nil num:nil html-style:nil
#+PANDOC_OPTIONS: reference-doc:reference.docx
#+END_EXAMPLE
** Environment Options
#+BEGIN_EXAMPLE emacs-lisp
;; On Windows
(if (eq system-type 'windows-nt)
    (setq load-path (cons "~/elisp/org-6.34c/lisp" load-path))
  (set-default-font "-outline-Consolas-normal-r-normal-normal-14-97-96-96-c-*-iso8859-1")
  )

;; On Linux
(if (eq system-type 'gnu/linux)
    (setq load-path (cons "~/elisp/org-current/lisp" load-path))
  (set-default-font "Inconsolata-11")
  )
#+END_EXAMPLE
** Personal Information
#+BEGIN_SRC emacs-lisp
(setq user-full-name "Sean Avery")
(setq user-mail-address "seanavery@gmx.net")
#+END_SRC
* Debug
#+BEGIN_SRC emacs-lisp
(setq debug-on-error t)
#+END_SRC

* Bootstrap
** Windows Variables
#+BEGIN_SRC emacs-lisp
(if (eq system-type 'windows-nt)
(setq org_notes (concat (getenv "HOME") "/Dropbox/zettelkasten/"))
(setq org_agenda (concat (getenv "HOME") "/Dropbox/zettelkasten/task.org"))
(setq org_bib (concat (getenv "HOME") "/Dropbox/zettelkasten/bibnotes.org"))
(setq org_export (concat (getenv "HOME") "/Dropbox/org-export/"))
(setq zot_bib (concat (getenv "HOME") "/Dropbox/zettelkasten/MeineBibliothek.bib")))
#+END_SRC

** Linux Variables
#+BEGIN_SRC emacs-lisp
(if (eq system-type 'gnu/linux)
(setq org_notes (concat (getenv "mnt") "/../../../mnt/c/Users/averh/Dropbox/zettelkasten/")))
(setq org_agenda (concat (getenv "mnt") "/../../../mnt/c/Users/averh/Dropbox/zettelkasten/task.org"))
(setq org_bib (concat (getenv "mnt") "/../../../mnt/c/Users/averh/Dropbox/zettelkasten/bibnotes.org"))
(setq org_export (concat (getenv "mnt") "/../../../mnt/c/Users/averh/Dropbox/org-export/"))
(setq zot_bib (concat (getenv "mnt") "/../../../mnt/c/Users/averh/Dropbox/zettelkasten/MeineBibliothek.bib"))
#+END_SRC

** UTF8 Everywhere
#+BEGIN_SRC emacs-lisp
(setq-default buffer-file-coding-system 'utf-8-unix)
(set-terminal-coding-system 'utf-8)
(set-language-environment 'utf-8)
(set-keyboard-coding-system 'utf-8)
(prefer-coding-system 'utf-8)
(setq locale-coding-system 'utf-8)
(set-default-coding-systems 'utf-8)
(set-terminal-coding-system 'utf-8)
#+END_SRC

** Initial Buffer
#+BEGIN_SRC emacs-lisp
(if (eq system-type 'windows-nt)
(setq initial-buffer-choice "~/Dropbox/zettelkasten/start-index.org"))
#+END_SRC

#+BEGIN_SRC emacs-lisp
(if (eq system-type 'gnu/linux)
(setq initial-buffer-choice "/mnt/c/Users/averh/Dropbox/zettelkasten/start-index.org"))
#+END_SRC
** Frame Settings
#+BEGIN_SRC emacs-lisp
(menu-bar-mode -1)                       ; Minimal menu bar
(tool-bar-mode -1)                       ; No toolbar
(scroll-bar-mode -1)                     ; Disable scroll bars
#+END_SRC

#+BEGIN_SRC emacs-lisp
(setq default-frame-alist
      '((top . 70) (left . 10)
        (width . 100) (height . 50)))

(setq initial-frame-alist '
      ((top . 70) (left . 10)
       (width . 100) (height . 50)))
#+END_SRC

** Mode Line Display
#+BEGIN_SRC emacs-lisp
(delete-selection-mode 1)                ; Replace selection when inserting text
(display-time-mode 1)                    ; Enable time in the mode-line
(display-battery-mode 1)                 ; It's nice to know how much power you have
(setq display-time-day-and-date t)       ; Display time and date
#+END_SRC

** Line Numbers
I prefer relative line numbers. Lets make it global.
#+BEGIN_SRC emacs-lisp
(setq display-line-numbers-type 'relative) ;
#+END_SRC
** File Movement
#+BEGIN_SRC emacs-lisp
(global-display-line-numbers-mode)       ; display line numbers
(setq line-move-visual t)                ; move around lines based on how they displayed
#+END_SRC

** Theme Settings
Doom has a couple of Themes. My favorite Gruvbox!
#+BEGIN_SRC emacs-lisp
(setq doom-theme 'doom-gruvbox)
(custom-set-faces! '(doom-modeline-evil-insert-state :weight bold :foreground "#339CDB"))
#+END_SRC

** Font Face
Setting fonts.
#+BEGIN_SRC emacs-lisp
doom-big-font (font-spec :family "JetBrains Mono" :size 36)                    ;
doom-variable-pitch-font (font-spec :family "Overpass" :size 24)               ;
;;doom-serif-font (font-spec :family "IBM Plex Mono" :weight 'light))          ;
;;(setq fancy-splash-image "~/image.png")                                      ;
#+END_SRC

** Separate Lines
[[https://yiufung.net/post/org-mode-hidden-gems-pt1/][Org-mode Hidden Gems - 01 Document Structure]]
#+BEGIN_SRC emacs-lisp
(setq org-cycle-separator-lines 0)
#+END_SRC

** Buffer Settings
Let Emacs initializes in Org-Mode
#+BEGIN_SRC emacs-lisp
;; Default Emacs mode on initializaiton
(setq-default major-mode 'org-mode)
#+End_SRC

Lets use some buffer names.
#+BEGIN_SRC emacs-lisp
(setq doom-fallback-buffer-name "$Doom"
      +doom-dashboard-name "$ Doom")
#+END_SRC

Ask which buffer to open after splitting the window.
#+BEGIN_SRC emacs-lisp
(setq evil-vsplit-window-right t
      evil-split-window-below t)

(defadvice! prompt-for-buffer (&rest _)
  :after '(evil-window-split evil-window-vsplit)
  (+ivy/switch-buffer))

(setq +ivy-buffer-preview t)
#+END_SRC

** Browser Settings!!!
#+BEGIN_SRC emacs-lisp
(setq
 browse-url-browser-function 'browse-url-generic
 browse-url-generic-program "firefox")
(setq browse-url-browser-function
      'browse-url-default-windows-browser)  ;win10
#+END_SRC

** Evil Settings
Some evil stuff
#+BEGIN_SRC emacs-lisp
(setq evil-move-cursor-back nil)
(setq evil-escape-key-sequence "jk"
      evil-escape-delay 0.63
      evil-escape-unordered-key-sequence t)
#+END_SRC

** Undo Settings
#+BEGIN_SRC emacs-lisp
(setq undo-limit 80000000                         ; Raise undo-limit to 80Mb
      evil-want-fine-undo t                       ; By default while in insert all changes are one big blob. Be more granular
      auto-save-default t                         ; Nobody likes to loose work, I certainly don't
      inhibit-compacting-font-caches t)           ; When there are lots of glyphs, keep them in memory
#+END_SRC

** Other Settings
#+BEGIN_SRC emacs-lisp
(setq org-catch-invisible-edits 'show-and-error)  ; Avoid inadvertent text edit in invisible area
(setq-default
 delete-by-moving-to-trash t                      ; Delete files to trash
 tab-width 4                                      ; Set width for tabs
 uniquify-buffer-name-style 'forward              ; Uniquify buffer names
 window-combination-resize t                      ; take new window space from all other windows (not just current)
 x-stretch-cursor t                               ; Stretch cursor to the glyph width
 global-subword-mode 1)                           ; Iterate through CamelCase words
#+END_SRC

** Windows Shell
#+BEGIN_SRC emacs-lisp
(if (eq system-type 'windows-nt)
(defun run-bash ()
      (interactive)
      (let ((shell-file-name "C:\\Program Files\\Git\\bin\\bash.exe"))
            (shell "*bash*")))

(defun run-cmdexe ()
      (interactive)
      (let ((shell-file-name "cmd.exe"))
            (shell "*cmd.exe*")))

(defun run-powershell ()
  "Run powershell"
  (interactive)
  (async-shell-command "c:/windows/system32/WindowsPowerShell/v1.0/powershell.exe -Command -"
               nil
               nil)))
#+END_SRC

#+BEGIN_SRC emacs-lisp
(if (eq system-type 'windows-nt)
  (setq w32-allow-system-shell t) ; enables cmd.exe as shell
  (setq save-interprogram-paste-before-kill 1 ; stop killing my clipboard, plz
        ; ghostscript on windows
        ; see https://www.emacswiki.org/emacs/docviewmode for details
        doc-view-ghostscript-program "~/Applications/msys2/mingw64/bin/gswin32c.exe"
        ; set curl location
        request-curl "~/Applications/msys2/mingw64/bin/curl.exe")
  (cond ((executable-find "aspell") ; spell-checking
         (setq ispell-program-name "aspell")
         (setq ispell-extra-args '("--sug-mode=ultra" "--lang=en_us")))))
#+END_SRC

** Windows System Variables
#+BEGIN_SRC emacs-lisp
(if (eq system-type 'windows-nt)
(setenv  "PATH" (concat
                ;; "c:/Windows/System32" ";"
                 "c:/Windows/Microsoft.NET/Framework/v4.0.30319" ";"
                 "C:\\Windows\\Microsoft.NET\\Framework\\v4.0.30319" ";"
                 ;; Unix tools
                 "C:\\Program Files\\Git\\usr\\bin" ";"
                 ;; User binary files
                 "C:\\User\\arch\\bin" ";"
                 ;; Mono Installation.
                 "c:\\Program Files\\Mono\\bin" ";"
                 "c:\\Program Files\\Mono\\lib\\mono\\4.5" ";"
                 (getenv "PATH")
         )))
#+END_SRC

** Windows Pdf-Tools & MySys64
[[https://gist.github.com/justbur/afb5f605def2376eeffc643f7b7daae9][Install pdf-tools on windows with msys2 · GitHub]]
#+BEGIN_SRC emacs-lisp
(setenv "PATH" (concat "C:\\msys64\\mingw64\\bin;" (getenv "PATH")))
(prefer-coding-system 'utf-8-unix)
#+END_SRC

** Windows Search Everything
#+BEGIN_SRC emacs-lisp
(when (eq system-type 'windows-nt)
(with-eval-after-load 'counsel
    (defun counsel-locate-cmd-es (input)
      "Return a shell command based on INPUT."
      (counsel-require-program "es.exe")
      (format "es.exe -r %s"
              (counsel--elisp-to-pcre
               (ivy--regex input t))))))
#+END_SRC

* Keybindings
#+BEGIN_EXAMPLE emacs-lisp
(map! :leader
:prefix "k"
:desc "amread-mode" "a" #'amread-mode
:desc "amread-start" "s" #'amread-start
:desc "amread-stop" "q" #'amread-stop
)

(map! :leader
      :desc "New journal entry"
      "l j j" #'org-journal-new-entry)
#+END_EXAMPLE
#+BEGIN_SRC emacs-lisp
(map! :leader
      (:prefix-map ("a" . "applications")
       (:prefix ("a" . "amread")
        :desc "amread-mode" "a" #'amread-mode
        :desc "amread-start" "s" #'amread-start
        :desc "amread-stop" "q" #'amread-stop
        :desc "New journal entry" "j" #'org-journal-new-entry
        :desc "Search journal entry" "s" #'org-journal-search
        )
        (:prefix ("t" . "terminal")
        :desc "bash" "b" #'run-cmdexe
        :desc "cmd" "c" #'run-bash
        )
        (:prefix ("k" . "kill-ring")
        :desc "cmd" "k" #'helm-show-kill-ring
        )
       ))
#+END_SRC

* Functions
** List current Miner-Modes
#+BEGIN_SRC emacs-lisp
(defun saa/current-minor-modes ()
  "Return the list of minor modes enabled in the current buffer."
  (interactive)
  (delq nil
	(mapcar (lambda (mode)
		  (if (and (boundp mode) (symbol-value mode))
		      mode))
		minor-mode-list)))
#+END_SRC

** Insert Today’s Date
#+BEGIN_SRC emacs-lisp
(require 'ts)

(defun saa-today (&optional arg)
"Insert today's date.

A prefix ARG specifies how many days to move;
negative means previous day.

If region selected, parse region as today's date pivot."
  (interactive "P")
  (let ((date (if (use-region-p)
                  (ts-parse (buffer-substring-no-properties (region-beginning) (region-end)))
                (ts-now)))
        (arg (or arg 0)))
    (if (use-region-p)
        (delete-region (region-beginning) (region-end)))
    (insert (ts-format "%A, %B %e, %Y" (ts-adjust 'day arg date)))))
#+END_SRC

** Insert Current Filename
#+BEGIN_SRC emacs-lisp
(defun saa/insert-filename-as-heading ()
  "Take current filename (word separated by dash) as heading."
  (interactive)
  (insert
   (capitalize
    (replace-regexp-in-string "-" " " (file-name-sans-extension (buffer-name))))))
#+END_SRC

** Rename Current Buffer-File
#+BEGIN_SRC emacs-lisp
(defun saa/rename-current-buffer-file ()
  "Renames current buffer and file it is visiting."
  (interactive)
  (let ((name (buffer-name))
        (filename (buffer-file-name)))
    (if (not (and filename (file-exists-p filename)))
        (error "Buffer '%s' is not visiting a file!" name)
      (let ((new-name (read-file-name "New name: " filename)))
        (if (get-buffer new-name)
            (error "A buffer named '%s' already exists!" new-name)
          (rename-file filename new-name 1)
          (rename-buffer new-name)
          (set-visited-file-name new-name)
          (set-buffer-modified-p nil)
          (message "File '%s' successfully renamed to '%s'"
                   name (file-name-nondirectory new-name)))))))

(global-set-key (kbd "C-x C-r") 'rename-current-buffer-file)
#+END_SRC

** Get String from File
Muss noch getestet werden.
#+BEGIN_SRC emacs-lisp
(defun zp/get-string-from-file (file-path)
  "Read file content from FILE-PATH."
  (with-temp-buffer
    (insert-file-contents file-path)
    (buffer-string)))
#+END_SRC

** Flyspell with Whitespace-Mode
#+BEGIN_SRC emacs-lisp
(defun saa/flyspell-and-whitespace-mode ()
  "Toggle `flyspell-mode' and `whitespace-mode'."
  (interactive)
  (if (derived-mode-p 'prog-mode)
      (flyspell-prog-mode)
    (flyspell-mode)
    (when flyspell-mode
      (flyspell-buffer)))
  (whitespace-mode 'toggle))

(global-set-key (kbd "<f9>") #'saa/flyspell-and-whitespace-mode)
#+END_SRC


** Create-Date/Update-Date
#+BEGIN_SRC emacs-lisp
    (autoload 'enable-paredit-mode "paredit" "Turn on pseudo-structural editing of Lisp code." t)
  (defun zp/org-find-time-file-property (property &optional anywhere)
    "Return the position of the time file PROPERTY if it exists.
When ANYWHERE is non-nil, search beyond the preamble."
    (save-excursion
      (goto-char (point-min))
      (let ((first-heading
             (save-excursion
               (re-search-forward org-outline-regexp-bol nil t))))
        (when (re-search-forward (format "^#\\+%s:" property)
                                 (if anywhere nil first-heading)
                                 t)
          (point)))))

  (defun zp/org-has-time-file-property-p (property &optional anywhere)
    "Return the position of time file PROPERTY if it is defined.
As a special case, return -1 if the time file PROPERTY exists but
is not defined."
    (when-let ((pos (zp/org-find-time-file-property property anywhere)))
      (save-excursion
        (goto-char pos)
        (if (and (looking-at-p " ")
                 (progn (forward-char)
                        (org-at-timestamp-p 'lax)))
            pos
          -1))))

  (defun zp/org-set-time-file-property (property &optional anywhere pos)
    "Set the time file PROPERTY in the preamble.
When ANYWHERE is non-nil, search beyond the preamble.
If the position of the file PROPERTY has already been computed,
it can be passed in POS."
    (when-let ((pos (or pos
                        (zp/org-find-time-file-property property))))
      (save-excursion
        (goto-char pos)
        (if (looking-at-p " ")
            (forward-char)
          (insert " "))
        (delete-region (point) (line-end-position))
        (let* ((now (format-time-string "[%Y-%m-%d %a %H:%M]")))
          (insert now)))))

  (defun zp/org-set-last-modified ()
    "Update the LAST_MODIFIED file property in the preamble."
    (when (derived-mode-p 'org-mode)
      (zp/org-set-time-file-property "LAST_MODIFIED")))
#+END_SRC

** Kill line without Kill-Ring
#+BEGIN_SRC emacs-lisp
(defun saa-delete-line ()
  "Delete from current position to end of line without pushing to `kill-ring'."
  (interactive)
  (delete-region (point) (line-end-position)))

(defun saa-delete-whole-line ()
  "Delete whole line without pushing to kill-ring."
  (interactive)
  (delete-region (line-beginning-position) (line-end-position)))

(defun saa-smart-delete-line ()
  "Kill to the end of the line and kill whole line on the next call."
  (interactive)
  (let ((orig-point (point)))
    (move-end-of-line 1)
    (if (= orig-point (point))
        (aza-delete-whole-line)
      (goto-char orig-point)
      (aza-delete-line))))
#+END_SRC

** Remind Me
#+BEGIN_SRC emacs-lisp
(defun play-reminder-sound ()
  (start-process "" nil "mpv" coin-work-medium-sound))

(defun saa/remind-me ()
  "Notify with a sound after certain time"
  (interactive)
  (let ((time (read-string "Time (min|sec): " "10 min")))
    (message "I will remind you after %s" time)
    (run-at-time time nil #'play-reminder-sound)))
#+END_SRC

** Playing Multimedia Files
#+BEGIN_SRC emacs-lisp
(defun start-mpv (path &optional playlist-p)
  "Start mpv with specified arguments"
  (let* ((default-cmd "mpv --force-window")
         (cmd (if playlist-p
                  (s-append " --loop-playlist --playlist=" default-cmd)
                (s-append " --loop " default-cmd))))
    (call-process-shell-command (s-concat cmd (shell-quote-argument path)) nil 0)))

(defun mpv ()
  "Play a file in current line"
  (interactive)
  (start-mpv (dired-get-filename)))

(defun mpv-dir ()
  "Play all multimedia files in current directory"
  (interactive)
  (start-mpv default-directory))

(defun mpv-playlist ()
  "Play a playlist in current line"
  (interactive)
  (start-mpv (dired-get-filename) t))
#+END_SRC

** TS
#+BEGIN_SRC emacs-lisp
(load! "~/../../../mnt/c/Users/averh/Dropbox/emacs-packages/ts/ts.el")
#+END_SRC

#+BEGIN_SRC emacs-lisp
(use-package! ts)
#+END_SRC

** Get A Day Name From A Date Time
#+BEGIN_SRC emacs-lisp
(defun saa/what-day ()
  "Show day name from spesific time"
  (interactive)
  (let ((date (read-string "Date: "))
        (month (read-string "Month: "))
        (year (read-string "Year: " (number-to-string (ts-year (ts-now))))))
    (message (ts-day-name (ts-parse (s-join " " (list date month year)))))))
#+END_SRC

** Switch Window
#+BEGIN_SRC emacs-lisp
(defun saa/switch-windows ()
  "Rotate your windows"
  (interactive)
  (cond ((not (> (count-windows)1))
         (message "You can't rotate a single window!"))
        (t
         (setq i 1)
         (setq numWindows (count-windows))
         (while  (< i numWindows)
           (let* (
                  (w1 (elt (window-list) i))
                  (w2 (elt (window-list) (+ (% i numWindows) 1)))

                  (b1 (window-buffer w1))
                  (b2 (window-buffer w2))

                  (s1 (window-start w1))
                  (s2 (window-start w2))
                  )
             (set-window-buffer w1  b2)
             (set-window-buffer w2 b1)
             (set-window-start w1 s2)
             (set-window-start w2 s1)
             (setq i (1+ i)))))))
#+END_SRC

** Join Lines
#+BEGIN_SRC emacs-lisp
(global-set-key (kbd "M-ö")
            (lambda ()
                  (interactive)
                  (join-line -1)))
#+END_SRC

* Server
This starts the Emacs server when .emacs gets loaded
#+BEGIN_SRC emacs-lisp
    (require 'server)
    (if (not (server-running-p)) (server-start))
#+END_SRC
This changes C-x C-c to just hide Emacs until the next
#+BEGIN_SRC emacs-lisp
    ;;  time you use it.  We rebind C-M-c to be the command to
    ;;  really kill Emacs.
    (defun my-done ()
      "Exit server buffers and hide the main Emacs window"
      (interactive)
      (server-edit)
      (make-frame-invisible nil t))
    (global-set-key (kbd "C-x C-c") 'my-done)
    (global-set-key (kbd "C-M-c") 'save-buffers-kill-emacs)
#+END_SRC

* Packages Windows
** (e,c)Emacs-Sql-Lisp
#+BEGIN_SRC emacs-lisp
(if (eq system-type 'windows-nt)
(use-package emacsql-sqlite3
  :custom (emacsql-sqlite-executable-path "C:/sqlite/sqlite3.exe")))
#+END_SRC

** (e,c)ISpell
On Windows:
    Install scoop
    scoop install msys2
    pacman -Syu aspell aspell-en
#+BEGIN_SRC emacs-lisp
(if (eq system-type 'windows-nt)
(setq ispell-program-name "C:/Users/averh/.doom.d/packages/aspell.cmd"))
#+END_SRC

#+BEGIN_EXAMPLE emacs-lisp
(when (boundp 'ispell-program-path)
  (add-to-list 'exec-path ispell-program-path))
(setq ispell-program-name "aspell")
       (add-to-list 'exec-path "C:/hunspell/bin/")

(setq ispell-program-name (locate-file "hunspell"
    exec-path exec-suffixes 'file-executable-p))

(setq ispell-local-dictionary-alist '(
       (nil
           "[[:alpha:]]"
           "[^[:alpha:]]"
           "[']"
           t
           ("-d" "en_US" "-p" "C:\\hunspell\\share\\hunspell\\personal.en")
           nil
           iso-8859-1)

       ("american"
           "[[:alpha:]]"
           "[^[:alpha:]]"
           "[']"
           t
           ("-d" "en_US" "-p" "C:\\hunspell\\share\\hunspell\\personal.en")
           nil
           iso-8859-1)
       ("deutsch"
           "[[:alpha:]ÄÖÜéäöüß]"
           "[^[:alpha:]ÄÖÜéäöüß]"
           "[']"
           t
           ("-d" "de_DE_frami" "-p"
"C:\\hunspell\\share\\hunspell\\personal.de")
           nil
           iso-8859-1)
       ("francais"
           "[[:alpha:]ÀÂÇÈÉÊËÎÏÔÙÛÜàâçèéêëîïôùûü]"
           "[^[:alpha:]ÀÂÇÈÉÊËÎÏÔÙÛÜàâçèéêëîïôùûü]"
           "[-']"
           t
           ("-d" "fr-classique" "-p"
"C:\\hunspell\\share\\hunspell\\personal.fr")
           nil
           utf-8)
        ))

#+END_EXAMPLE

* Packages Installed
** (e,c)Ace-Window
This package aims to take the speed and predictability of windmove and pack it into a single key binding, similar to other-window. Just assign ace-window to a short key binding, as switching windows is a common task. I suggest M-o, as it's short and not bound to anything important in the default Emacs. \\
[[https://github.com/abo-abo/ace-window/tree/7003c88cd9cad58dc35c7cd13ebc61c355fb5be7][GitHub - abo-abo/ace-window at 7003c88cd9cad58dc35c7cd13ebc61c355fb5be7]]
#+BEGIN_SRC emacs-lisp
(global-set-key (kbd "M-o") 'ace-window)
#+END_SRC

** (e,c)Amread
This is a minor mode helping user speed-reading.
#+BEGIN_SRC emacs-lisp
(require 'amread-mode)
(use-package amread-mode
  :ensure t
  :commands (amread-mode))
(map! :leader
:prefix "j"
:desc "amread-mode" "a" #'amread-mode
:desc "amread-start" "s" #'amread-start
:desc "amread-stop" "q" #'amread-stop
)
#+END_SRC

** (e,c)Auto-Dim-Other-Buffers
Auto-Dim-Buffer
#+BEGIN_SRC emacs-lisp
(require 'auto-dim-other-buffers)
(add-hook 'after-init-hook (lambda ()
  (when (fboundp 'auto-dim-other-buffers-mode)
    (auto-dim-other-buffers-mode t))))
#+END_SRC

** (e,c)Auto-Dictionary
https://github.com/nschum/auto-dictionary-mode
#+BEGIN_SRC emacs-lisp
(require 'auto-dictionary)
(add-hook 'flyspell-mode-hook (lambda () (auto-dictionary-mode 1)))

#+END_SRC
** (i,c)Company
It’s nice to have completions almost all the time, in my opinion. Key strokes are just waiting to be saved!
Now, the improvements from precident are mostly from remembering history, so let’s improve that memory.
#+BEGIN_SRC emacs-lisp
(after! company
  (setq company-idle-delay 0.5
        company-minimum-prefix-length 2)
  (setq company-show-numbers t)
(add-hook 'evil-normal-state-entry-hook #'company-abort)) ;; make aborting less annoying.
;===========================================
(setq-default history-length 1000)
(setq-default prescient-history-length 1000)
#+END_SRC

** (e,c)Command-Log
Show event history and command history of some or all buffers \\
[[https://github.com/lewang/command-log-mode][GitHub - lewang/command-log-mode: log commands to buffer]]
#+BEGIN_SRC emacs-lisp

config.el
;;;;(global-set-key (kbd "C-x p i") 'org-cliplink)

(global-undo-tree-mode)

(setq user-full-name "Sean Avery"
      user-mail-address "seanavery@gmx.net")

(setq doom-font (font-spec :family "JetBrains Mono" :size 14))
(setq doom-theme 'doom-gruvbox)
;;(setq fancy-splash-image "~/image.png")

(setq org-support-shift-select t)
(setq display-line-numbers-type 'relative)
(setq evil-move-cursor-back nil)                 ; Escape
(setq evil-escape-key-sequence "jk"              ; Escape
      evil-escape-delay 0.3)                     ; Escape
(custom-set-faces! '(doom-modeline-evil-insert-state :weight bold :foreground "#339CDB"))

(setq-default
 delete-by-moving-to-trash t                      ; Delete files to trash
 tab-width 4                                      ; Set width for tabs
 uniquify-buffer-name-style 'forward              ; Uniquify buffer names
 window-combination-resize t                      ; take new window space from all other windows (not just current)
 x-stretch-cursor t)                              ; Stretch cursor to the glyph width

(setq undo-limit 80000000                         ; Raise undo-limit to 80Mb
      evil-want-fine-undo t                       ; By default while in insert all changes are one big blob. Be more granular
      auto-save-default t                         ; Nobody likes to loose work, I certainly don't
      inhibit-compacting-font-caches t)           ; When there are lots of glyphs, keep them in memory

(delete-selection-mode 1)                         ; Replace selection when inserting text
(display-time-mode 1)                             ; Enable time in the mode-line
(display-battery-mode 1)                          ; On laptops it's nice to know how much power you have
(global-subword-mode 1)                           ; Iterate through CamelCase words

(setq-default major-mode 'org-mode)

;;(if (eq initial-window-system 'x)                 ; if started by emacs command or desktop file
;;    (toggle-frame-maximized)
;;  (toggle-frame-fullscreen))

(setq evil-vsplit-window-right t
      evil-split-window-below t)

(defadvice! prompt-for-buffer (&rest _)
  :after '(evil-window-split evil-window-vsplit)
  (+ivy/switch-buffer))

(setq +ivy-buffer-preview t)

(move-text-default-bindings)

(use-package deft
:after org
:bind
("C-c n d" . deft)
:custom
(deft-recursive t)
(deft-use-filter-string-for-filename t)
(deft-default-extension "org")
(deft-extensions '("txt" "tex" "org"))
(deft-directory "~/Dropbox/zettelkasten"))

(setq org-directory "~/org/")

(use-package org-roam-protocol)

(defun ha/org-return (&optional ignore)
  "Add new list item, heading or table row with RET.
A double return on an empty element deletes it.
Use a prefix arg to get regular RET. "
  (interactive "P")
  (if ignore
      (org-return)
    (cond
     ;; Open links like usual
     ((eq 'link (car (org-element-context)))
      (org-return))
     ;; lists end with two blank lines, so we need to make sure we are also not
     ;; at the beginning of a line to avoid a loop where a new entry gets
     ;; created with only one blank line.
     ((and (org-in-item-p) (not (bolp)))
      (if (org-element-property :contents-begin (org-element-context))
          (org-insert-heading)
        (beginning-of-line)
        (setf (buffer-substring
               (line-beginning-position) (line-end-position)) "")
        (org-return)))
     ((org-at-heading-p)
      (if (not (string= "" (org-element-property :title (org-element-context))))
          (progn (org-end-of-meta-data)
                 (org-insert-heading))
        (beginning-of-line)
        (setf (buffer-substring
               (line-beginning-position) (line-end-position)) "")))
     ((org-at-table-p)
      (if (-any?
           (lambda (x) (not (string= "" x)))
           (nth
            (- (org-table-current-dline) 1)
            (org-table-to-lisp)))
          (org-return)
        ;; empty row
        (beginning-of-line)
        (setf (buffer-substring
               (line-beginning-position) (line-end-position)) "")
        (org-return)))
     (t
      (org-return)))))

(define-key org-mode-map (kbd "RET")  #'ha/org-return)

(setq org-roam-directory "~/Dropbox/zettelkasten/")
(setq org-roam-buffer-position 'right)
(setq org-roam-buffer-width 0.4)
(setq org-roam-index-file "index.org")
(setq org-roam-completion-system 'ivy)(after! org-roam
(map! :leader
:prefix "n"
:desc "org-roam" "l" #'org-roam
:desc "org-roam-insert" "i" #'org-roam-insert
:desc "org-roam-switch-to-buffer" "b" #'org-roam-switch-to-buffer
:desc "org-roam-find-file" "f" #'org-roam-find-file
:desc "org-roam-show-graph" "g" #'org-roam-show-graph
:desc "org-roam-insert" "i" #'org-roam-insert
:desc "org-roam-capture" "c" #'org-roam-capture
:desc "org-roam-find-directory" "d" #'org-roam-find-directory
:desc "org-roam-company" "k" #'company-org-roam
:desc "org-roam-find-ref" "x" #'org-roam-find-ref
:desc "org-roam-graph" "y" #'org-roam-graph
:desc "org-roam-jump-to-index" "j" #'org-roam-jump-to-index
))

(require 'company-org-roam)
(use-package company-org-roam
:when (featurep! :completion company)
:after org-roam
:config
(set-company-backend! 'org-mode '(company-org-roam company-yasnippet company-dabbrev)))

(use-package! org-roam-server)

;;(use-package! company-org-roam
;;  :config
;;  (push 'company-org-roam company-backends))

(defun zp/org-find-time-file-property (property &optional anywhere)
    "Return the position of the time file PROPERTY if it exists.
When ANYWHERE is non-nil, search beyond the preamble."
    (save-excursion
      (goto-char (point-min))
      (let ((first-heading
             (save-excursion
               (re-search-forward org-outline-regexp-bol nil t))))
        (when (re-search-forward (format "^#\\+%s:" property)
                                 (if anywhere nil first-heading)
                                 t)
          (point)))))

  (defun zp/org-has-time-file-property-p (property &optional anywhere)
    "Return the position of time file PROPERTY if it is defined.
As a special case, return -1 if the time file PROPERTY exists but
is not defined."
    (when-let ((pos (zp/org-find-time-file-property property anywhere)))
      (save-excursion
        (goto-char pos)
        (if (and (looking-at-p " ")
                 (progn (forward-char)
                        (org-at-timestamp-p 'lax)))
            pos
          -1))))

  (defun zp/org-set-time-file-property (property &optional anywhere pos)
    "Set the time file PROPERTY in the preamble.
When ANYWHERE is non-nil, search beyond the preamble.
If the position of the file PROPERTY has already been computed,
it can be passed in POS."
    (when-let ((pos (or pos
                        (zp/org-find-time-file-property property))))
      (save-excursion
        (goto-char pos)
        (if (looking-at-p " ")
            (forward-char)
          (insert " "))
        (delete-region (point) (line-end-position))
        (let* ((now (format-time-string "[%Y-%m-%d %a %H:%M]")))
          (insert now)))))

  (defun zp/org-set-last-modified ()
    "Update the LAST_MODIFIED file property in the preamble."
    (when (derived-mode-p 'org-mode)
      (zp/org-set-time-file-property "LAST_MODIFIED")))

(setq org-roam-capture-templates
'(
  ("d" "default" plain (function org-roam--capture-get-point)
   "%?"
    :file-name "%(format-time-string \"%Y%m%d%H%M%S-${slug}\" (current-time) t)"
    :head "#+TITLE: ${title}\n#+Author: Sean Averhoff\n#+CREATED: %U\n#+LAST_MODIFIED: %U\n- tags :: \n----------\n"
        :unnarrowed t)

  ("n" "note" plain (function org-roam--capture-get-point)
  "%?"
    :file-name  "%(format-time-string \"%Y%m%d%H%M%S-${slug}\" (current-time) t)"
    :head "#+TITLE: ${title}\n#+Author: Sean Averhoff\n#+CREATED: %U\n#+LAST_MODIFIED: %U\n- tags :: \n----------\n"
    :unnarrowed t)

  ("p" "private" plain (function org-roam--capture-get-point)
  "%?"
    :file-name "private-${slug}"
    :head "#+TITLE: ${title}\n#+Author: ${author-or-editor}"
    :unnarrowed t)))

(setq +mu4e-backend 'offlineimap)
